overlord/snapstate/autorefresh_gating.go

```diff
@@ -32,6 +35,249 @@ import (
+		return snapst.LastRefresh, nil
+	}
+	inf, err := snapst.CurrentInfo()
+	if err != nil {
+		return nil, err
+	}
+	fst, err := os.Stat(inf.MountFile())
+	if err != nil {
+		return nil, err
+	}
+	t := fst.ModTime()
+	return &t, nil
+}
+
+func refreshGating(st *state.State) (map[string]map[string]*holdInfo, error) {
+	// affecting snap -> gating snap(s) -> first-held/hold-until time
+	var gating map[string]map[string]*holdInfo
+	err := st.Get("refresh-gating", &gating)
+	if err != nil && err != state.ErrNoState {
+		return nil, fmt.Errorf("internal error: cannot get refresh-gating: %v", err)
+	}
+	if err == state.ErrNoState {
+		return make(map[string]map[string]*holdInfo), nil
+	}
+	return gating, nil
+}
+
+// HoldError contains the details of snaps that cannot to be held.
+type HoldError struct {
+	SnapsInError map[string]error
+}
+
+func (h *HoldError) Error() string {
+	l := []string{""}
+	for _, e := range h.SnapsInError {
+		l = append(l, e.Error())
+	}
+	return fmt.Sprintf("cannot hold some snaps:%s", strings.Join(l, "\n - "))
+}
+
+// HoldRefresh marks affectingSnaps as held for refresh for up to holdTime.
+// HoldTime of zero denotes maximum allowed hold time.
+// Holding may fail for only some snaps in which case HoldError is returned and
+// it contains the details of failed ones.
+func HoldRefresh(st *state.State, gatingSnap string, holdTime time.Time, affectingSnaps ...string) error {
+	gating, err := refreshGating(st)
+	if err != nil {
+		return err
+	}
+	herr := &HoldError{
+		SnapsInError: make(map[string]error),
+	}
+	now := timeNow()
+	for _, affecting := range affectingSnaps {
+		refreshed, err := lastRefreshed(st, affecting)
+		if err != nil {
+			return err
+		}
+		if now.Sub(*refreshed) > maxPostponement {
+			herr.SnapsInError[affecting] = fmt.Errorf("cannot hold the refresh of snap %q, maximum postponement time exceeded", affecting)
+			continue
+		}
+		if _, ok := gating[affecting]; !ok {
+			gating[affecting] = make(map[string]*holdInfo)
+		}
+		hold, ok := gating[affecting][gatingSnap]
+		if !ok {
+			hold = &holdInfo{
+				FirstHeld: now,
+			}
+		}
+		// determine the maximum time the snap can be held by this gating snap.
+		var maxHoldDur time.Duration
+		if affecting == gatingSnap {
+			maxHoldDur = maxPostponement
+		} else {
+			maxHoldDur = maxOtherHoldDuration
+		}
+		var holdDur time.Duration
+		// explicit hold time requested
+		if !holdTime.IsZero() {
+			holdDur = holdTime.Sub(hold.FirstHeld)
+			if holdDur > maxHoldDur {
+				herr.SnapsInError[affecting] = fmt.Errorf("requested holding time %s exceeds maximum holding time for snap %q", holdTime, affecting)
+				continue
+			}
+		} else {
+			holdDur = maxHoldDur
+		}
+		// HoldUntil is relative to the time the snap was first held, so in case
+		// of consecutive refreshes and hold attempts this cannot go beyond
+		// maximum allowed hold time.
+		hold.HoldUntil = hold.FirstHeld.Add(holdDur)
+		gating[affecting][gatingSnap] = hold
```

Comment:
when will resetGating be called, I'm asking because this is adding/changing entries but we might also entries to drop because a snap doesn't have a refresh anymore?

Reply:
resetGating is going to be called early in auto-refresh, probably as soon as we have current refresh-candidates.